gordon/interfaces.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2017 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from zope.interface import Attribute
+from zope.interface import Interface
+
+
+class IEventMessage(Interface):
+    """A discrete unit of work for Gordon to process.
+
+    Gordon expects plugins to return or accept objects that implement this
+    interface in order to route them to other plugins, and handle retries
+    or cleanup in case of errors.
+    """
+    msg_id = Attribute('Unique message id')
+    phase = Attribute('Current phase of message')
+    data = Attribute('DNS change data')
+    history_log = Attribute('Log of actions performed on the message')
+
+    def __init__(msg_id, data, phase=None):
+        """Initialize an EventMessage.
+
+        Args:
+            msg_id (str): Unique message identifier.
+            data (dict): Data required to update DNS records.
+            phase (str): Current phase.
+        """
+
+    def append_to_history(entry):
+        """Append entry to the EventMessage's history_log.
+
+        Args:
+            entry (str): Information to append to log.
+        """
+
+
+class IGenericPlugin(Interface):
+    """Do not implement this interface directly.
+
+    Use IEventClient, IEnricherClient, or IPublisherClient instead.
+    """
+    phase = Attribute('Plugin phase')
+
+    def __init__(config, success_channel, error_channel, loop, metrics=None):
+        """Initialize an EventClient object.
+
+        Args:
+            config (dict): plugin-specific configuration
+            success_channel (asyncio.Queue): a sink for successfully created
+                IEventMessages.
+            error_channel (asyncio.Queue): a sink for invalid messages.
```

Comment:
I think "successfully created IEventMessage" and "invalid messages" is not the correct wording. Not sure what would be correct, but something like "sink for messages when plugin client successfully processes message" and "...failed to process message"